src/primitive_file_win.cc

```diff
@@ -154,32 +153,113 @@ PRIMITIVE(open) {
     // with open (eg a pipe, a socket, a directory).  We forbid this because
     // these file descriptors can block, and this API does not support
     // blocking.
-    INVALID_ARGUMENT;
+    if (strcmpi(R"(\\.\NUL)", pathname) != 0)
+      INVALID_ARGUMENT;
   }
   closer.clear();
   return Smi::from(fd);
 }
 
-class Directory {
+class Directory : public SimpleResource {
  public:
   TAG(Directory);
-  DIR* dir;
+  explicit Directory(SimpleResourceGroup* resource_group, const char* path) : SimpleResource(resource_group) {
+    snprintf(path_, MAX_PATH, "%s\\*", path);
+  }
+
+  const char* path() { return path_; }
+  WIN32_FIND_DATA* find_file_data() { return &find_file_data_; }
+  void set_dir_handle(HANDLE dir_handle) { dir_handle_ = dir_handle; }
+  HANDLE dir_handle() { return dir_handle_; }
+  bool done() const { return done_; }
+  void set_done(bool done) { done_ = done; }
+
+ private:
+  char path_[MAX_PATH]{};
+  WIN32_FIND_DATA find_file_data_{};
+  HANDLE dir_handle_ = INVALID_HANDLE_VALUE;
+  bool done_ = false;
 };
 
 PRIMITIVE(opendir) {
   UNIMPLEMENTED_PRIMITIVE;
 }
 
 PRIMITIVE(opendir2) {
-  UNIMPLEMENTED_PRIMITIVE;
+  ARGS(SimpleResourceGroup, group, cstring, pathname);
+  char path[MAX_PATH];
+  auto error = get_relative_path(process, pathname, path);
+  if (error) return error;
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  auto directory = _new Directory(group, path);
+  if (!directory) MALLOC_FAILED;
+
+  HANDLE dir_handle = FindFirstFile(directory->path(), directory->find_file_data());
+  if (dir_handle == INVALID_HANDLE_VALUE) {
+    if (GetLastError() == ERROR_NO_MORE_FILES) directory->set_done(true);
```

Comment:
I prefer adding { } to balance the else part. If find it hard to read this version.